### ISSUE TITLE
[LoopVectorizer] More efficient vector runtime checks

### ISSUE THREAD

**Issue by davemgreen** [llvm/llvm-project#53590]
Right now in llvm we generate "full loop bound" checks for required memchecks in the vectorizer:
https://godbolt.org/z/9Ma7qx8vd

Especially for the nested loop case, the runtime checks become a meaningful percentage of the total runtime. They check, for two loads and a store in a simple loop:
`or (and (icmp ult (d, s1+n), icmp ult (s1, d+n)),  and (icmp ult (d, s2+n), icmp ult (s2, d+n)))`

We can attempt to optimize those runtime checks in the backend a bit, but will never get to optimal unless we change the way they are generated. For vector runtime checks, we do not need to check that the entire range of the array accessed by the loop does not overlap. We only need to check the bounds for each loop iteration. Not only does this allow more precise bounds for when the vector code can be executed, it is also simpler for the backend to produce checks for.

For this simple case, we only need to check that `d - s <u VF * IC * TypeSize` (or something close to that, minus off by one errors). That can be selected in AArch64 as a `sub x, d, s; cmp x, #C; br cc`. We could also allow the s==d case, but that might require an extra add, as GCCs codegen contains.

The performance differences I measured were as high as 15% with the correct types/cpus/loop iteration counts. And this can come up in quite a lot of places, wherever vectorization require memory checks.

**Comment by fhahn** [comment]
I looked into this and put up a first version of a patch (still a bit WIP) to improve the runtime checks: https://reviews.llvm.org/D119078.

Another thing that can help to reduce runtime overhead of checks is running the 'extra vectorizer passes' which (`-mllvm -extra-vectorizer-passes`). Not in the case you mentioned, but I've seen a couple of cases where it can help a lot, especially if runtime checks can be moved out of loops.

**Comment by davemgreen** [comment]
Oh, that's great. Thanks Florian. I had a very (very) quick patch that did roughly the same thing, minus any resemblance of correctness checks. It was just for checking whether it was going to give some benefit. I wasn't looking forward to figuring out all the edge cases, but from the look of your patch it doesn't look a bad as I had first imagined.

So long as all the edge cases work (like reverses, interleaving, offsets, etc), it sounds like a great start, that should cover a lot of the more common cases.

**Comment by fhahn** [comment]
> So long as all the edge cases work (like reverses, interleaving, offsets, etc), it sounds like a great start, that should cover a lot of the more common cases.

The initial patch doesn't yet support all cases where we could use the difference checks instead of the full overlap checks, but the supported cases should be handled correctly. For example, interleaving or inductions with step != 1/-1 are not supported yet. Support for those can be added as follow-up, once we are confident the basic support works as expected.

The test suite also doesn't really cover much of the various edge-cases, so I put up a patch to cover more interesting scenarios: https://reviews.llvm.org/D119121. There are plenty of additional interesting test cases of course, but it's a start.

**Comment by davemgreen** [comment]
That sounds good. By "work" I just meant "works correctly", not necessarily "is optimal". Improvements can be made as we go.

**Comment by vfdff** [comment]
I try a debug with the 2nd simplified test on https://github.com/llvm/llvm-project/issues/53398 base on https://reviews.llvm.org/D119078, and the vectorization is still inhibit by the checking of memory conflicts,
```
10345   if (!LVL.canVectorize(EnableVPlanNativePath)) {
10346     LLVM_DEBUG(dbgs() << "LV: Not vectorizing: Cannot prove legality.\n");
10347     Hints.emitRemarkWithHints();
10348     return false;
10349   }
```
so it does not have a chance to try create a runtime checking later (while gcc done vectorization with  runtime checking).

**Comment by fhahn** [comment]
> > Another thing that can help to reduce runtime overhead of checks is running the 'extra vectorizer passes' which (-mllvm -extra-vectorizer-passes). Not in the case you mentioned, but I've seen a couple of cases where it can help a lot, especially if runtime checks can be moved out of loops.
> 
> This was implemented in https://reviews.llvm.org/D115052, right? But sadly, user needs to explicitly enable it with --extra-vectorizer-passes.. Maybe it makes sense to default extra-vectorizer-passes to true given you only run these passes on vectorized loops after D115052 so the impact is minimal.

Yes this landed. Last time I measured the impact it was still quite notable (see below), but I don't think the sequence of extra vectorizer passes has been fine-tuned so there should be potential to shave off a bit more


```
NewPM-O3: +1.43%
NewPM-ReleaseThinLTO: +1.00%
NewPM-ReleaseLTO-g: +1.58%
```
